### PR TITLE
Fnr

### PR DESCRIPTION
**Pull request by toolslive**
Note: 
- this will do absolutely nothing until the manifest serialization version is 2 in the NSM.
- enabling this (along with infrastructure to trigger it at the right moment) will be done in a different pr.

### CODE REVIEW

File: ocaml/src/osd.ml [hunk 1]
```diff
@@ -45,9 +45,16 @@ module Osd_namespace_state = struct
 end
 
 module Error = Asd_protocol.Protocol.Error
+
+type apply_result' = ((key * string option) list) [@@deriving show]
+
 type apply_result =
```
Comment (domsj) [inline ocaml/src/osd.ml]: hmm, looking at this type we may want use Prelude.Result instead?

File: ocaml/src/tools/prelude.ml [hunk 2]
```diff
@@ -191,6 +191,16 @@ module List = struct
 
   let find_index f l = find_index' f 0 l
 
+  let assoc' eq k kvs =
+    let rec inner = function
+      | [] -> raise Not_found
```
Comment (toolslive) [inline ocaml/src/tools/prelude.ml]: ok.

File: ocaml/src/alba_client_upload.ml [hunk 3]
```diff
@@ -107,7 +113,19 @@ let upload_chunk
       ~osds
       ~(fragment_cache : Fragment_cache.cache)
       ~cache_on_write
-      ~upload_slack
+      ~upload_slack :
+      ((Alba_statistics.Statistics.fragment_upload
+        * (Albamgr_protocol.Protocol.Osd.id option
+           * Asd_protocol.checksum
+           * bytes option
+           * string option
```
Comment (toolslive) [inline ocaml/src/alba_client_upload.ml]: ok.

File: ocaml/src/asd_server.ml [hunk 4]
```diff
@@ -1222,12 +1222,32 @@ let execute_update : type req res.
...
+                       (fun acc ku ->
+                         match ku with
+                         | (key, `Set (value, _)) ->
+                            begin
+                              let open Value in
+                              let e =
+                                match value with
+                                | Direct v -> key, None
+                                | OnFs (fnr, _) ->
+                                   let fnr_s = serialize Llio.int64_to fnr in
```
Comment (toolslive) [inline ocaml/src/asd_server.ml]: ok. It could also be a varint in rocksdb. (it's viral)

File: ocaml/src/osd.ml (hunk 1 above)
Comment (toolslive) [inline ocaml/src/osd.ml]: what about Pervasives.result ?

Comment (domsj) [inline ocaml/src/osd.ml]: That's actually the same type if I'm not mistaken?

File: ocaml/src/nsm_client.ml [hunk 5]
```diff
@@ -169,32 +171,55 @@ class client (nsm_host_client : Nsm_host_client.basic_client) namespace_id =
...
-            | Nsm_exn (Unknown_operation,_) ->
-               supports_update_manifest2 := Some false;
-               old_call ()
-            | exn -> Lwt.fail exn
-           )
-
+      | Some 1 -> old_call ()
+      | None ->
+         begin
+           let call = function
```
Comment (domsj) [inline ocaml/src/nsm_client.ml]: some repetition here matching on 3, 2, 1 that could be avoided

**Pull request merged**